### PR TITLE
Update system addressbook card only when there was a change

### PR DESCRIPTION
**Pull request by MorrisJobke**
Due to our old and new hook system the card dav backend listens to old and new hooks. This triggers this code multiple times and always causes an update. With this change we only do the database update when there is actually a change in the card. This reduces the updates to the database in exchange for a read request (which should be faster).

That is one approach. Another one would be to cache this directly in PHP. Let me do this as well and then we can see what is better suited.

cc @rullzer

### CODE REVIEW

**Comment by MorrisJobke** [comment]
> That is one approach. Another one would be to cache this directly in PHP. Let me do this as well and then we can see what is better suited.

The other one is #21733 which @rullzer and I would favor.

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 1]
```diff
@@ -694,6 +694,18 @@ public function updateCard($addressBookId, $cardUri, $cardData) {
...
+		$result = $query->select($query->func()->count())
+			->from($this->dbCardsTable)
+			->where($query->expr()->eq('uri', $query->createNamedParameter($cardUri)))
+			->andWhere($query->expr()->eq('addressbookid', $query->createNamedParameter($addressBookId)))
+			->andWhere($query->expr()->eq('etag', $query->createNamedParameter($etag)))
+			->execute();
+		if ((int)$result->fetchColumn() === 1) {
+			return '"' . $etag . '"';
+		}
+
```
Comment (nickvergessen) [inline apps/dav/lib/CardDAV/CardDavBackend.php:708]:
```suggestion

$result->closeCursor();

```

**Review by nickvergessen — CHANGES_REQUESTED** [review]
See comment, you need to reset the query object

**Comment by MorrisJobke** [comment]
> See comment, you need to reset the query object

Ah ... didn't knew. Most likely we will not go for this approach as there is a simpler and still nearly as effective approach: #21733

**Pull request closed without merging**